prov/util/src/util_coll.c

```diff
@@ -983,3 +983,46 @@ void ofi_coll_handle_xfer_comp(uint64_t tag, void *ctx)
 	util_ep = container_of(xfer_item->hdr.coll_op->mc->ep, struct util_ep, ep_fid);
 	util_coll_op_progress_work(util_ep, xfer_item->hdr.coll_op);
 }
+
+int ofi_query_collective(struct fid_domain *domain, struct fi_collective_attr *attr,
+			 enum fi_datatype datatype, enum fi_op op, uint64_t flags)
+{
+	int ret;
+
+	if (!attr || attr->mode != 0)
+		return -FI_EINVAL;
+
+	switch (attr->coll) {
+	case FI_BARRIER:
+		if ((enum fi_collective_op) op == FI_BARRIER && datatype)
+			ret = FI_SUCCESS;
+		else
+			ret = -FI_EINVAL;
+		break;
+	case FI_ALLREDUCE:
+		if (FI_MIN <= op && FI_BXOR >= op)
+			ret = fi_query_atomic(domain, datatype, op, &attr->datatype_attr, flags);
+		else
+			return -FI_ENOSYS;
+		break;
+	case FI_BROADCAST:
+	case FI_ALLTOALL:
+	case FI_ALLGATHER:
+	case FI_REDUCE_SCATTER:
+	case FI_REDUCE:
+	case FI_SCATTER:
+	case FI_GATHER:
+	default:
+		return -FI_ENOSYS;
+	}
+
+	if (ret)
+		return ret;
+
+	// assuming unlimited resources, the number of members
+	// is only limited by our ability to 'address' them in the
+	// collective tag, which restricts the rank to 31 bits.
+	attr->max_members = ~(0x80000000);
```

Comment:
@shefty I wasn't sure what other factors would effect the max_members. Am i missing something?

Reply:
Limits may be imposed by hardware, or we may restrict the number of members based on the type of collective and the amount of memory needed to support that number.  With a software based implementation, we may be restricted by the size of our local receive buffer allocated to a collective in those cases where each peer needs their own dedicated space.  For example, if we split an 8k buffer into 64-byte integers, then for some operations we may only be able to suport 128 members.

Reply:
that makes sense.  In the currently supported collective operations, however, i don't think we have these restrictions as all of them are reductions and transfer only the current value of the reduction.  i will update the comment here to include your points.